[Metricbeat] Create elb metricset using light weight module

kaiyan-sheng opened the pull request:
This PR is to use elastic/beats#12465 to create `elb` metricset for aws. `elb` metricset collects metrics from aws cloudwatch with namespace AWS/ELB using `aws` module `cloudwatch` metricset. 

There is also a pre defined dashboard added in this PR:
<img width="1009" alt="Screen Shot 2019-08-01 at 11 08 42 AM" src="https://user-images.githubusercontent.com/14081635/62312968-c3fa8480-b44c-11e9-87c0-33236e2507cf.png">

closes https://github.com/elastic/beats/issues/11701

kaiyan-sheng commented on x-pack/metricbeat/include/list.go:
```diff
@@ -15,6 +15,7 @@ import (
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/s3_daily_storage"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/s3_request"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/sqs"
+	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws_light"
```
@jsoriano question about the list.go, should `_ "github.com/elastic/beats/x-pack/metricbeat/module/cockroachdb/status"` be included in this list.go as well?

kaiyan-sheng commented on x-pack/metricbeat/include/list.go:
```diff
@@ -15,6 +15,7 @@ import (
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/s3_request"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/sqs"
+	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws_light"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/cockroachdb"
```
@jsoriano question about the list.go, should _ "github.com/elastic/beats/x-pack/metricbeat/module/cockroachdb/status" be included in this list.go as well?

jsoriano commented on metricbeat/docs/modules/cockroachdb/status.asciidoc:
```diff
@@ -15,3 +15,9 @@ include::../../../../x-pack/metricbeat/module/cockroachdb/status/_meta/docs.asci
+[source,json]
+----
+include::../../../../x-pack/metricbeat/module/cockroachdb/status/_meta/data.json[]
+----
```
Thanks!

jsoriano commented on x-pack/metricbeat/magefile.go:
```diff
@@ -90,7 +90,7 @@ func TestPackages() error {
 
 		// To be increased or removed when more light modules are added
-		devtools.MinModules(1))
+		devtools.MinModules(2))
```
:+1:

jsoriano commented on x-pack/metricbeat/module/aws/elb/_meta/data.json:
```diff
@@ -0,0 +1,33 @@
+{
+    "@timestamp": "2017-10-12T08:05:34.853Z",
+    "aws": {
+        "elb": {
```
I wonder if we should place all cloudwatch metrics under `aws.cloudwatch.metrics` as we do with prometheus that we store all metrics under `prometheus.metrics`, so we can reuse dynamic mappings.
There are other fields to distinguish services, as `aws.cloudwatch.namespace` and `event.dataset`.

jsoriano commented on x-pack/metricbeat/module/aws/elb/elb.go:
```diff
@@ -0,0 +1,22 @@
+	// To be moved to some kind of helper
+	os.Setenv("BEAT_STRICT_PERMS", "false")
+	mb.Registry.SetSecondarySource(xpackmb.NewLightModulesSource("../../../module"))
+}
```
This should be in a test file (move to `elb_test.go` for example), if not we are overriding strict perms and secondary source in the final binary.

jsoriano commented on x-pack/metricbeat/module/aws/elb/elb_integration_test.go:
```diff
@@ -0,0 +1,26 @@
+	metricSet := mbtest.NewReportingMetricSetV2Error(t, config)
+	if err := mbtest.WriteEventsReporterV2Error(metricSet, t, "/"); err != nil {
+		t.Fatal("write", err)
+	}
```
This can be implemented now as:
```
metricSet := mbtest.NewFetcher(t, config)
metricSet.WriteEvents(t, "/")
```
So we don't need to modify tests if we change the interface of the metricset (#11998)

kaiyan-sheng replied:
aha great! thanks for pointing it out!

jsoriano commented on x-pack/metricbeat/module/aws/module.yml:
```diff
@@ -0,0 +1,9 @@
+  - rds
+  - s3_daily_storage
+  - s3_request
+  - sqs
```
In principle only light metricsets should appear here.

jsoriano commented on x-pack/metricbeat/include/list.go:
```diff
@@ -11,6 +11,7 @@ import (
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/cloudwatch"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/ec2"
+	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/elb"
```
In principle light metricsets shouldn't appear here as they don't have code, this one appears here probably because of the `elb.go` file, that I think should be moved to `elb_test.go`.

kaiyan-sheng commented:
> This is looking good, are you planning to create a dashboard? 🙂

@jsoriano Yes, after adding a dashboard I will make this PR in review (officially 😬)!

jsoriano commented on x-pack/metricbeat/module/aws/elb/_meta/data.json:
```diff
@@ -26,7 +58,6 @@
     },
     "service": {
-        "name": "cloudwatch",
         "type": "cloudwatch"
```
I wonder why `service.type` is still set to `cloudwatch` :thinking:

kaiyan-sheng replied:
Yeah I realized all aws metricsets have the wrong `service.type` so I left it and was thinking to create a separate PR to fix it. But I guess adding in this PR is fine.

exekias commented on x-pack/metricbeat/module/aws/elb/_meta/data.json:
```diff
@@ -0,0 +1,63 @@
+            "namespace": "AWS/ELB"
+        },
+        "metrics": {
+            "EstimatedALBActiveConnectionCount": {
```
I guess default config can now be updated after merging https://github.com/elastic/beats/pull/12840? I hope we can be picky about the statistics we retrieve, limiting it only to the ones that make sense / we plan to use

exekias commented on x-pack/metricbeat/module/aws/elb/manifest.yml:
```diff
@@ -0,0 +1,7 @@
+  metricset: cloudwatch
+  defaults:
+    metrics:
+      - namespace: AWS/ELB
```
for my previous comment, that would mean we define the metrics we want here :)

kaiyan-sheng replied:
Yep, I added 
```
defaults:
    metrics:
      - namespace: AWS/ELB
        statistic: ["Average"]
        tags.resource_type_filter: elasticloadbalancing
```

exekias replied:
This is a good description of all available metrics, I see there are different statistics recommended depending on the metric, this is a good opportunity to leverage your latest updates here! https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-cloudwatch-metrics.html

I'm curious, I see many of these are not present in `data.json`, do you know why?

kaiyan-sheng replied:
Good point! I will change it 😬 to adapt different statistic methods. A lot of the metrics have no value at the time when `data.json` is generated. For example, here is a screenshot from cloudwatch for ELB latency metric:
<img width="1400" alt="Screen Shot 2019-08-01 at 8 42 04 AM" src="https://user-images.githubusercontent.com/14081635/62302916-6ceab480-b438-11e9-954c-0cf3f0cd29e8.png">

kaiyan-sheng commented:
@jsoriano I added a dashboard for ELB and moved this PR to in review. I waited for a while to try to get better looking data but so far no luck by using the test aws env 😬

jsoriano commented on x-pack/metricbeat/module/aws/_meta/kibana/7/dashboard/Metricbeat-aws-elb-overview.json:
```diff
@@ -0,0 +1,995 @@
+            }
+          }
+        },
+        "title": "AWS ELB HTTP 5XX Errors",
```
Visualization titles should have the metricset as suffix
```suggestion
        "title": "HTTP 5XX Errors [AWS ELB]",
```

kaiyan-sheng replied:
👍 I will fix all visualizations. Thanks!

jsoriano replied:
Oh sorry, I should have said `ELB HTTP 5XX Errors [Metricbeat AWS]`
